File: src/calibration.py
"""Per-user neutral head-pose calibration."""

from __future__ import annotations

from dataclasses import dataclass, field

from head_pose import HeadPose

DURATION_SEC = 1.5


@dataclass
class NeutralPoseCalibrator:
    """Collects a short rest-pose sample and subtracts it from later poses."""

    duration_sec: float = DURATION_SEC
    _yaw_samples: list[float] = field(default_factory=list)
    _pitch_samples: list[float] = field(default_factory=list)
    _collected_sec: float = 0.0
    _last_update_time: float | None = None
    _baseline_yaw: float | None = None
    _baseline_pitch: float | None = None
    _active: bool = False

    @property
    def is_active(self) -> bool:
        return self._active

    @property
    def is_complete(self) -> bool:
        return self._baseline_yaw is not None and self._baseline_pitch is not None

    @property
    def baseline_yaw(self) -> float | None:
        return self._baseline_yaw

    @property
    def baseline_pitch(self) -> float | None:
        return self._baseline_pitch

    @property
    def progress(self) -> float:
        if self.is_complete:
            return 1.0
        if self.duration_sec <= 0:
            return 1.0
        return min(1.0, self._collected_sec / self.duration_sec)

    def start(self) -> None:
        self._yaw_samples.clear()
        self._pitch_samples.clear()
        self._collected_sec = 0.0
        self._last_update_time = None
        self._baseline_yaw = None
        self._baseline_pitch = None
        self._active = True

    def update(self, pose: HeadPose | None, now: float) -> None:
        if not self._active or self.is_complete:
            return

        if pose is None:
            self._last_update_time = None
            return

        if self._last_update_time is not None:
            delta = now - self._last_update_time
            if delta > 0:
                self._collected_sec += delta
        self._last_update_time = now

        self._yaw_samples.append(pose.yaw_deg)
        self._pitch_samples.append(pose.pitch_deg)

        if self._collected_sec >= self.duration_sec:
            self._baseline_yaw = sum(self._yaw_samples) / len(self._yaw_samples)
            self._baseline_pitch = sum(self._pitch_samples) / len(self._pitch_samples)
            self._active = False

    def apply(self, pose: HeadPose) -> HeadPose:
        if not self.is_complete:
            return pose

        return HeadPose(
            yaw_deg=pose.yaw_deg - self._baseline_yaw,
            pitch_deg=pose.pitch_deg - self._baseline_pitch,
            roll_deg=pose.roll_deg,
        )

File: lib/calibration.ts
import type { HeadPose } from "@/lib/head-pose";

export const DURATION_SEC = 1.5;

export class NeutralPoseCalibrator {
  durationSec: number;
  private yawSamples: number[] = [];
  private pitchSamples: number[] = [];
  private collectedSec = 0;
  private lastUpdateTime: number | null = null;
  private _baselineYaw: number | null = null;
  private _baselinePitch: number | null = null;
  private active = false;

  constructor(durationSec = DURATION_SEC) {
    this.durationSec = durationSec;
  }

  get isActive(): boolean {
    return this.active;
  }

  get isComplete(): boolean {
    return this._baselineYaw !== null && this._baselinePitch !== null;
  }

  get baselineYaw(): number | null {
    return this._baselineYaw;
  }

  get baselinePitch(): number | null {
    return this._baselinePitch;
  }

  start(): void {
    this.yawSamples = [];
    this.pitchSamples = [];
    this.collectedSec = 0;
    this.lastUpdateTime = null;
    this._baselineYaw = null;
    this._baselinePitch = null;
    this.active = true;
  }

  update(pose: HeadPose | null, now: number): void {
    if (!this.active || this.isComplete) {
      return;
    }

    if (pose === null) {
      this.lastUpdateTime = null;
      return;
    }

    if (this.lastUpdateTime !== null) {
      const delta = now - this.lastUpdateTime;
      if (delta > 0) {
        this.collectedSec += delta;
      }
    }
    this.lastUpdateTime = now;

    this.yawSamples.push(pose.yawDeg);
    this.pitchSamples.push(pose.pitchDeg);

    if (this.collectedSec >= this.durationSec && this.yawSamples.length > 0) {
      this._baselineYaw =
        this.yawSamples.reduce((sum, value) => sum + value, 0) / this.yawSamples.length;
      this._baselinePitch =
        this.pitchSamples.reduce((sum, value) => sum + value, 0) / this.pitchSamples.length;
      this.active = false;
    }
  }

  apply(pose: HeadPose): HeadPose {
    if (!this.isComplete || this._baselineYaw === null || this._baselinePitch === null) {
      return pose;
    }

    return {
      yawDeg: pose.yawDeg - this._baselineYaw,
      pitchDeg: pose.pitchDeg - this._baselinePitch,
      rollDeg: pose.rollDeg,
    };
  }
}
